### DISCUSSION TITLE
Why Microsoft.EntityFrameworkCore.Design is not included in WebUI?

### DISCUSSION THREAD

**Discussion by GianlucaLocri** [jasontaylordev/CleanArchitecture#690] [Q&A] [answered]
I'm using your awesome template, but I cannot perform migrations as per sample code on the repo readme.
(I've already disabled in memory database)

The error is:
"Your startup project 'CleanArchitecture.WebUI' doesn't reference Microsoft.EntityFrameworkCore.Design.

Should I go ahead by adding the package?
I feel this is against the code clean architecture.

**Comment by lwestfall** [comment, marked as answer]
(Not the owner of this repo) What I do instead is implement a custom IDesignTimeDbContextFactory in the Infrastructure project, and omit the `--startup-project` argument. I have two separate codebases that this works really well for. Here's a sample (had to remove a few proprietary details, so there may be an error or two).

Doing this will eliminate the need to reference Microsoft.EntityFrameworkCore.Design in your WebUI project.

(PS if someone smarter than me sees pitfalls here - please lmk!)

```cs
public class MyAwesomeDbContextFactory : IDesignTimeDbContextFactory<MyAwesomeDbContext>
{
    public MyAwesomeDbContext CreateDbContext(string[] args)
    {
        // sample migration add command (run from src/Infrastructure): dotnet ef migrations add <MigrationName> -c MyAwesomeDbContext -o \Persistence\Migrations\ -- Development

        var env = args[0];

        var jsonFile = "";

        if (env == "Production")
        {
            jsonFile = "appsettings.json";
        }
        else if (env == "Development")
        {
            jsonFile = "appsettings.Development.json";
        }
        else
        {
            throw new NotImplementedException("No appsettings.json for environment named " + env);
        }

        var configBuilder = new ConfigurationBuilder()
            .SetBasePath(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location))
            .AddJsonFile(jsonFile)
            .AddEnvironmentVariables();

        var optionsBuilder = new DbContextOptionsBuilder<MyAwesomeDbContext>();
        var config = configBuilder.Build();

        optionsBuilder
            .UseMyAwesomeDbContext(config)

        return new MyAwesomeDbContext(optionsBuilder.Options, config);
    }
}
```

**Reply by GianlucaLocri** [reply]
Well, this seems a good solution!

**Reply by GianlucaLocri** [reply]
I'm trying your solution but I've a doubt:
Are you using the appsetting file from the WebUI project?
It seems that you have created a new appsetting for the infrastructure project. Am I right?

**Reply by lwestfall** [reply]
Yes, sorry I forgot about that. I moved my appsettings.json files from WebUI under src/AppSettings. If you do this you'll similarly need `.SetBasePath(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location))` in your WebUI Startup / Program as well. The benefit here is that you don't need to duplicate the files between Infrastructure and WebUI.

Let me know if you have any other questions!

**Comment by prince272** [comment]
I have another trick. Just keep the appsettings.json inside the WebUI and use this code inside the Infrastructure,

        private static IConfigurationRoot BuildConfiguration()
        {
            static string GetBasePath()
            {
                string currentDirectory = Directory.GetCurrentDirectory();
                string baseDirectory = currentDirectory;

                // Split the path into its directory components.
                string[] directories = currentDirectory.Split(Path.DirectorySeparatorChar);

                // Check each component in reverse order.
                for (int i = directories.Length - 1; i >= 0; i--)
                {
                    baseDirectory = Path.Combine(string.Join(Path.DirectorySeparatorChar.ToString(), directories.Take(i + 1)), "MyBoilerplate.Server");
                    if (Directory.Exists(baseDirectory)) break;
                }

                return baseDirectory;
            }

            var environmentName = Environment.GetEnvironmentVariable("ASPNETCORE_ENVIRONMENT");
            var builder = new ConfigurationBuilder()
                         .SetBasePath(GetBasePath())
                         .AddJsonFile("appsettings.json", optional: true, reloadOnChange: true)
                         .AddJsonFile($"appsettings.{environmentName}.json", optional: true);

            return builder.Build();
        }
